### PR TITLE
Include view hierarchy description with screenshots in bug reports

### PR DESCRIPTION
**Pull request by NickEntin**

### CODE REVIEW

File: Aardvark/ARKEmailBugReporter.m [hunk 1]
```diff
@@ -37,6 +37,82 @@ @interface ARKInvisibleView : UIView
 @end
 
 
+@interface CALayer (HierarchyDescription)
+
+- (NSString *)_recursiveLayerHierarchyDescriptionWithIndentationLevel:(int)indentationLevel;
```
Comment (dfed) [inline Aardvark/ARKEmailBugReporter.m]: we should prefix the names of extensions with `ARK_` or `_ARK_` to prevent potential collisions with Apple's code.

File: Aardvark/ARKEmailBugReporter.m [hunk 2]
```diff
@@ -37,6 +37,82 @@ @interface ARKInvisibleView : UIView
...
+#pragma mark - Private Methods
+
+- (NSString *)_recursiveViewHierarchyDescriptionWithIndentationLevel:(int)indentationLevel;
+{
+    NSMutableString *const mutableDescription = [NSMutableString new];
+    for (int i = 0 ; i < indentationLevel ; i++) {
+        [mutableDescription appendString:@"   | "];
+    }
+    
+    [mutableDescription appendFormat:@"%@\n", [self description]];
```
Comment (dfed) [inline Aardvark/ARKEmailBugReporter.m]: curious: do you need to call `description` here?

File: Aardvark/ARKEmailBugReporter.m [hunk 3]
```diff
@@ -37,6 +37,82 @@ @interface ARKInvisibleView : UIView
...
+    
+    // Each subview's layer is also a sublayer, but we should only include it in the hierarchy once (with the subview).
+    NSMutableSet<CALayer *> *sublayersToSkip = [NSMutableSet<CALayer *> setWithCapacity:self.subviews.count];
```
Comment (dfed) [inline Aardvark/ARKEmailBugReporter.m]: supernit: `NSMutableSet<CALayer *> *const sublayersToSkip`

File: Aardvark/ARKEmailBugReporter.h [hunk 4]
```diff
@@ -68,6 +68,9 @@
 /// The window level for the email composer on iOS 7 or later. Defaults to UIWindowLevelStatusBar + 3.0.
 @property (nonatomic) UIWindowLevel emailComposeWindowLevel;
 
+/// If this bug report includes a screenshot, also attach a description of the view hierarchy. Defaults to YES.
+@property (nonatomic) BOOL attachesViewHierarchyWithScreenshot;
```
Comment (pwesten) [inline Aardvark/ARKEmailBugReporter.h]: ViewHierarchy**Description**?

File: Aardvark/ARKEmailBugReporter.m [hunk 5]
```diff
@@ -335,6 +420,14 @@ - (void)_createBugReportWithTitle:(NSString *)title
                             }
                         }
                         
+                        if (self.viewHierarchyDescription != nil) {
+                            NSData *viewHierarchyData = [self.viewHierarchyDescription dataUsingEncoding:NSUTF8StringEncoding];
```
Comment (dfed) [inline Aardvark/ARKEmailBugReporter.m]: supernit: `const`

File: Aardvark/ARKEmailBugReporter.m (hunk 1 above)
Comment (pwesten) [reply Aardvark/ARKEmailBugReporter.m]: +1

File: Aardvark/ARKEmailBugReporter.m [hunk 6]
```diff
@@ -335,6 +420,14 @@ - (void)_createBugReportWithTitle:(NSString *)title
...
+                            if (viewHierarchyData.length) {
```
Comment (dfed) [inline Aardvark/ARKEmailBugReporter.m]: `.length > 0` please. Implicit 0 comparisons are 🤢

File: Aardvark/ARKEmailBugReporter.m [hunk 7]
```diff
@@ -37,6 +37,82 @@ @interface ARKInvisibleView : UIView
...
+@interface CALayer (HierarchyDescription)
+
+- (NSString *)_recursiveLayerHierarchyDescriptionWithIndentationLevel:(int)indentationLevel;
+
+@end
+
+
+@implementation CALayer (HierarchyDescription)
+
+- (NSString *)_recursiveLayerHierarchyDescriptionWithIndentationLevel:(int)indentationLevel;
```
Comment (pwesten) [inline Aardvark/ARKEmailBugReporter.m]: Should probably be `NSUInteger` instead of `int`

File: Aardvark/ARKEmailBugReporter.m [hunk 8]
```diff
@@ -37,6 +37,82 @@ @interface ARKInvisibleView : UIView
...
+{
+    NSMutableString *const mutableDescription = [NSMutableString new];
+    for (int i = 0 ; i < indentationLevel ; i++) {
+        [mutableDescription appendString:@"   | "];
+    }
+    
+    [mutableDescription appendFormat:@"%@\n", [self description]];
+    
+    for (CALayer *sublayer in self.sublayers) {
+        [mutableDescription appendString:[sublayer _recursiveLayerHierarchyDescriptionWithIndentationLevel:(indentationLevel + 1)]];
```
Comment (pwesten) [inline Aardvark/ARKEmailBugReporter.m]: Totally doesn't matter but it would be a bit more efficient to pass the mutable string into the method as a parameter.

File: Aardvark/ARKEmailBugReporter.m [hunk 9]
```diff
@@ -37,6 +37,82 @@ @interface ARKInvisibleView : UIView
...
+    
+    return mutableDescription;
+}
+
+@end
+
+
+@interface UIView (HierarchyDescription)
+
+/// Re-implementation of `-[UIView recursiveDescription]` to avoid calling private API.
```
Comment (pwesten) [inline Aardvark/ARKEmailBugReporter.m]: The fact that this replicates SPI is somewhat orthogonal; I'd lead with a straightforward description of what it does, and maybe say parenthetically that it's like `recursiveDescription`. Also should have `ARK_` prefix.

File: Aardvark/ARKEmailBugReporter.m (hunk 2 above)
Comment (pwesten) [reply Aardvark/ARKEmailBugReporter.m]: Indeed you do not. :-)

Comment (NickEntin) [reply Aardvark/ARKEmailBugReporter.m]: Originally had it calling `debugDescription`, but changed it.  Will remove.

File: Aardvark/ARKEmailBugReporter.m (hunk 6 above)
Comment (NickEntin) [reply Aardvark/ARKEmailBugReporter.m]: This and non-const copied from the screenshot section directly above.  Will change both.  :)

**Comment by NickEntin** [comment]
In regards to adding a delegate for adding attachments, I think that would be a very valuable addition.  That said, I think this is something that would be useful in most cases, so I think it should be included by default.  I created a new issue to track adding that delegate: https://github.com/square/Aardvark/issues/55

File: Aardvark/ARKEmailBugReporter.m [hunk 10]
```diff
@@ -103,6 +206,17 @@ - (void)composeBugReportWithScreenshot:(BOOL)attachScreenshot;
         // Take a screenshot.
         ARKLogScreenshot();
         
+        if (self.attachesViewHierarchyDescriptionWithScreenshot) {
+            NSMutableString *mutableViewHierarchyDescription = [NSMutableString new];
+            for (UIWindow *window in [UIApplication sharedApplication].windows) {
+                NSMapTable<UIView *, UIViewController *> *const viewControllerMap = [NSMapTable<UIView *, UIViewController *> weakToWeakObjectsMapTable];
```
Comment (pwesten) [inline Aardvark/ARKEmailBugReporter.m]: I think strong-to-strong is actually better here.

**Comment by dfed** [comment]
Changes look good!
